daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ScriptF.scala

```diff
@@ -75,8 +74,44 @@ object ScriptF {
       _clients =
         _clients.copy(party_participants = _clients.party_participants + (party -> participant))
     }
-    def lookupChoice(id: Identifier, choice: Name): Either[String, Ast.TemplateChoiceSignature] =
-      compiledPackages.interface.lookupChoice(id, choice).map(_.choice).left.map(_.pretty)
+    def lookupChoice(
+        tmplId: Identifier,
+        ifaceId: Option[Identifier],
+        choice: Name,
+    ): Either[String, Ast.TemplateChoiceSignature] =
+      compiledPackages.interface.lookupChoice(tmplId, ifaceId, choice).left.map(_.pretty)
+
+    private[this] val archiveId: Identifier = Identifier.assertFromString(
+      "d14e08374fc7197d6a0de468c968ae8ba3aadbf9315476fd39071831f5923662:DA.Internal.Template:Archive"
+    )
+
+    // This is a workaround to infer the template/interface id where the choice is defined
+    // using the choice argument type. The inference should be done in daml.
+    def lookupChoiceByArgType(
+        tmplId: Identifier,
+        argTypeId: Identifier,
+    ): Either[String, Identifier] =
+      if (argTypeId == archiveId)
+        Right(tmplId)
+      else {
+        val Identifier(argPkg, QualifiedName(argMod, _)) = argTypeId
+        val choiceName = argTypeId.qualifiedName.name.segments.head
+        val argTyp = Ast.TTyCon(argTypeId)
+        compiledPackages.interface
+          .lookupModule(argPkg, argMod)
+          .left
+          .map(_.pretty)
+          .flatMap(mod =>
+            (mod.templates.view.mapValues(_.choices) ++
+              mod.interfaces.view.mapValues(_.fixedChoices))
+              .collectFirst {
```

Comment:
this also seems sensitive to choice clashes

Reply:
oh, I see it now, `_.argBinder._2 == argTyp` takes care of clashes since the _names_ might match but the _Identifier_ will only match with the choice we want. Nevermind :)